enhancements/etcd/protecting-etcd-quorum-during-control-plane-scaling.md

```diff
@@ -0,0 +1,317 @@
+
+* Automation of scaling operations on Machines
+* Horizontal scaling of the Etcd cluster and Control Plane
+* Providing these protection mechanisms when the Machine API is [not functional](#When-is-Machine-API-Functional)
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator of a managed OpenShift cluster, I want to be able to automate the scaling operations of Control Plane Machines without having to manually ensure the safety of the Etcd cluster.
+
+#### Story 2
+
+As a developer of OpenShift, I want to implement safety mechanisms that adhere to best practices for Etcd scaling operations to protect end users from potential quorum losses and data losses.
+
+#### Story 3
+
+As an end user of OpenShift, I want to be able to increase the size of my Control Plane without having to know the intracacies of Etcd and protecting its quorum.
+
+### Implementation Details/Notes/Constraints
+
+#### Requirements for Etcd safety
+
+* The number of voting members should equal the desired number of Control Plane Machines (and this should be an odd number)
+    * We may only deviate from this to add a replacement member
+    * Once the new member is added, we should remove old members as soon as possible to reduce the risk of degrading the cluster while having an even number of voting members
+* Existing voting members should not be removed until their replacements have been promoted to voting members
+    * By starting new Etcd members as [learners](https://etcd.io/docs/v3.3/learning/learner/#raft-learner), we can ensure that the new member is fully "caught-up" and promotable to a full voting member before we start the removal process of the old member
+    * This protects Etcd from potential inconsistencies in its data if the cluster were to have some interruption shortly after a new member joins the cluster
+    * This also ensures that we keep the full Etcd data on disk on a minimum of the desired number of Control Plane Machines at all times
+
+#### Protecting Etcd during scaling operations
+
+To ensure that the safety requirements described above are maintained during scaling operations, the Etcd Operator will manage the Etcd cluster in clusters with a [functional Machine API](#When-is-Machine-API-Functional), by leveraging [Machine Deletion Hooks](#What-are-Machine-Deletion-Hooks) to coordinate with the Machine API when it is, and isn't safe, to remove Machines with voting members of the Etcd cluster running on them.
+
+##### Overview of protecion mechanism
+
+To ensure the safety of the Etcd cluster quorum, the Etcd Operator will leverage a pre-drain [Machine Deletion Hooks](#What-are-Machine-Deletion-Hook) to prevent the removal of any Control Plane Machine hosting a voting member of the Etcd quorum.
+
+```yaml
+pre-drain.delete.hook.machine.cluster-api.x-k8s.io/etcd-quorum-member: openshift-etcd-operator
+```
+
+The Etcd Operator will apply the hook to a Machine resource once it identifies that the Machine hosts an Etcd quorum member.
```

Comment:
IIRC, we decided this would be the case so that the deletion protection only comes in once we know that machine needs proection. Thought thinking about it now, I think we should add the hook before we promote for the following reasons:
- If we promote before adding the hook, there's a small window before the protection kicks in
- There should be no harm in having the hook in place while the member is still a learner, when the etcd operator notices the machine is marked for deletion, it should remove the member from the etcd cluster (voting or learner) and then remove the hook, I think that should still be safe

I'll amend this and add context